In [4]:
import numpy as np
import math as mat

# We start with the uncertainties in the faint end slope

* This is done in one of two ways:
<ol>
<li> Pick a center value. Find sum of squared distances from other values to that center value. Take the square root and then divide by the number of other values. </li>
<li> Take all values. Find mean, and find sum of squared distances of all values from that mean. Take square root and divide by N-1 </li>
</ol>

In [11]:
#we will have 3 values: one for alpha = -1.25, one for alpha = -1.3, one for alpha = -1.35 (in that order)
# all use a lower bound of 10^-3
#center value is index 1
values_SR = [220.1,233.4,254.9]
values_LSS = [145.8,158.1,172.2]

In [12]:
# we start with method 1. We pick the center value as index 1, which was calculated with a slope of -1.3 
#and a bound of 10^-4 

def rms_center(center_ind,values):
    rms = 0.0
    for val in values:
        dist = val - values[center_ind]
        rms += dist**2
    
    rms = (rms/(len(values)-1))**0.5
    print('The uncertainty is {act:.1f}, or {rel:.2f}% of the center value'\
      .format(act = rms,rel=100*rms/values[center_ind]))
    
    return [rms,rms/values[center_ind]]

print('Lower bound and Alpha, Small Radii')
print('------------------------------ \n')
uncert_SR = rms_center(1,values_SR)
print('\n Lower bound and Alpha, Large Radii')
print('------------------------------ \n')
uncert_LSS = rms_center(1,values_LSS)

Lower bound and Alpha, Small Radii
------------------------------ 

The uncertainty is 17.9, or 7.66% of the center value

 Lower bound and Alpha, Large Radii
------------------------------ 

The uncertainty is 13.2, or 8.37% of the center value


In [13]:
# here's method 2. This just simply uses the mean.
def rms_mean(values):
    rms_2 = 0.0
    mean = np.mean(values)
    print(mean)

    for val in values:
        dist_sqrd = (val - mean)**2
        rms_2 += dist_sqrd

    rms_2 = (rms_2/(len(values)-1))**0.5

    print('The uncertainty is {act:.1f}, or {rel:.2f}% of the mean'\
          .format(act = rms_2,rel=100*rms_2/mean))
    
    return [rms_2,rms_2/mean]

uncert_mean_SR = rms_mean(values_SR)
uncert_mean_LSS = rms_mean(values_LSS)

print('\n' + str(100*uncert_mean_SR[0] / values_SR[1]))
print('\n' + str(100*uncert_mean_LSS[0] / values_LSS[1]))

236.13333333333333
The uncertainty is 17.6, or 7.44% of the mean
158.7
The uncertainty is 13.2, or 8.32% of the mean

7.523683437377703

8.355612469185571


* For small radii, we get an uncertainty of 17.6, which is 7.5% of 233
* For large radii, we get an uncertainty of 13.2, which is 8.4% of 158

# Now, we're going to find the uncertainty in the Leuthaud slope. 
This uses the same functions as above

**First, we check how different slopes compare to the data in Leuthaud.**

In [14]:
import pandas as pd
blue_line = pd.read_csv('./Data/LeuthBlueLine.csv',header=None)
leuth_points = pd.read_csv('./Data/LeuthLowLumPoints.csv',header = None)

#print(blue_line)
#print(leuth_points)

In [15]:
'''
import matplotlib.pyplot as plt
import random

def sloped_line(L,slope):
    norm = 1.05e7
    y = norm * L**(slope)
    return y

lums = np.arange(1e7,5e10,1e8)

leuth_line_lums = blue_line[0]
leuth_line_mPerMstar = blue_line[1]

#slope of -0.58
five_eight_fracs = sloped_line(lums,-0.58)
#slope of -0.56
five_six_fracs = sloped_line(lums,-0.56)
#slope of -0.54
five_four_fracs = sloped_line(lums,-0.54)
#slope of -0.52
five_two_fracs = sloped_line(lums,-0.52)
#slope of -0.50
five_oh_fracs = sloped_line(lums,-0.50)

funcs = [five_eight_fracs,five_six_fracs,five_four_fracs,five_two_fracs,five_oh_fracs]
names = ['-0.58','-0.56','-0.54','-0.52','-0.50']
colors = ['g','k','m','y','purple']

fig,axs = plt.subplots(nrows = 5,ncols = 1,figsize = (14,20))
for i in range(len(funcs)):
    axs[i].plot(leuth_line_lums,leuth_line_mPerMstar,label = 'Leuthaud Main Fit')
    axs[i].scatter(leuth_points[0],leuth_points[1],color = 'r',alpha = 0.5,label = 'Leuthaud Low Lum Points')

    axs[i].set_xscale('log',basex=10)
    axs[i].set_yscale('log',basey=10)
    
    axs[i].set_xlim(1e7,5e10)
    axs[i].set_ylim(30,5e3)

    axs[i].plot(lums,funcs[i],color = colors[i],label = 'slope = '+ names[i],alpha = 0.7)
    axs[i].legend(bbox_to_anchor=(1.05, 1))
    
    axs[i].set_xlabel('Stellar Mass (Solar Masses)')
    axs[i].set_xlabel('Total Mass / Stellar Mass')
    axs[i].set_title('Slope of ' + names[i],fontsize = 16)

fig.subplots_adjust(hspace = 0.7)

plt.savefig('./Figures/Different_Leuthaud_Slopes.png')
plt.show()'''

"\nimport matplotlib.pyplot as plt\nimport random\n\ndef sloped_line(L,slope):\n    norm = 1.05e7\n    y = norm * L**(slope)\n    return y\n\nlums = np.arange(1e7,5e10,1e8)\n\nleuth_line_lums = blue_line[0]\nleuth_line_mPerMstar = blue_line[1]\n\n#slope of -0.58\nfive_eight_fracs = sloped_line(lums,-0.58)\n#slope of -0.56\nfive_six_fracs = sloped_line(lums,-0.56)\n#slope of -0.54\nfive_four_fracs = sloped_line(lums,-0.54)\n#slope of -0.52\nfive_two_fracs = sloped_line(lums,-0.52)\n#slope of -0.50\nfive_oh_fracs = sloped_line(lums,-0.50)\n\nfuncs = [five_eight_fracs,five_six_fracs,five_four_fracs,five_two_fracs,five_oh_fracs]\nnames = ['-0.58','-0.56','-0.54','-0.52','-0.50']\ncolors = ['g','k','m','y','purple']\n\nfig,axs = plt.subplots(nrows = 5,ncols = 1,figsize = (14,20))\nfor i in range(len(funcs)):\n    axs[i].plot(leuth_line_lums,leuth_line_mPerMstar,label = 'Leuthaud Main Fit')\n    axs[i].scatter(leuth_points[0],leuth_points[1],color = 'r',alpha = 0.5,label = 'Leuthaud Low Lum 

In [17]:
# the arrays are for values as follows: first value is for a M/L proportional to L^(-0.56), then our center value of 
# L^(-0.54), then L^(-0.52)

leuth_vals_SR = [247.6,233.4,220.4]
leuth_vals_LSS = [167.5,158.1,149.4]

print('Leuthaud Slope, Small Radii')
print('------------------------------ \n')

uncert_leuth_SR = rms_mean(leuth_vals_SR)

print('\n Leuthaud Slope, Large Radii')
print('------------------------------ \n')

uncert_leuth_LSS = rms_mean(leuth_vals_LSS)

print('\n' + str(100*uncert_leuth_SR[0] / leuth_vals_SR[1]))
print('\n' + str(100*uncert_leuth_LSS[0] / leuth_vals_LSS[1]))

Leuthaud Slope, Small Radii
------------------------------ 

233.79999999999998
The uncertainty is 13.6, or 5.82% of the mean

 Leuthaud Slope, Large Radii
------------------------------ 

158.33333333333334
The uncertainty is 9.1, or 5.72% of the mean

5.828796507867133

5.725651931784277


* SR: 13.6, or 5.8% of 233
* LSS: 9.1, or 5.7% of 158

# And finally, we're going to do the uncertainty in galaxy morphology.

* **For small radii, we consider three cases**:
<ol>
    
    <li> Base Case: <ul> <li> Bright: 10% Sp, 90% E </li>
        <li>Faint: 20% Sp, 80% E </li> </ul> </li>
    
    <li> Case 2: <ul> <li> Bright: 10% Sp, 90% E </li>
        <li> Faint: 30% Sp, 70% E </li> </ul> </li>
        
    <li> Case 3: <ul> <li> Bright: 10% Sp, 90% E  </li>
	<li> Faint: 10% Sp, 90% E </li> </ul> </li>
</ol>

* **For Large Scale Structure, we consider 6 cases**:
<ol> 
    
    <li> Base Case: <ul> <li> Bright: 50% Sp, 50% E </li>
	<li> Faint: 80% Sp, 20% E </li> </ul> </li>
     
     <li> Case 2 (more bright Sp): <ul> <li> Bright: 70% Sp, 30% E </li>
         <li>Faint: 80% Sp, 20% E </li> </ul> </li>
     
     <li> Case 3 (more faint E): <ul> <li> Bright: 50% Sp, 50% E</li>
     <li>Faint: 70% Sp, 30% E </li> </ul> </li>
     
     <li> Case 4 (more bright Sp, more faint E): <ul> <li>Bright: 70% Sp, 30% E</li>
     <li>Faint: 70% Sp, 30% E</li> </ul> </li>
     
     <li> Case 5 (more faint Sp):  <ul> <li>Bright: 50% Sp, 50% E</li>
     <li>Faint: 90% Sp, 10% E</li> </ul> </li>
     
     <li> Case 6 (more faint Sp, more bright Sp): <ul> <li> Bright: 70% Sp, 30% E</li>
	<li>Faint: 90% Sp, 10% E</li> </ul> </li>
</ol>
    
  
+ The values in each array are listed in the same order

In [20]:
morph_vals_SR = [233.4,221.6,245.3]
morph_vals_LSS = [158.1,156.0,169.9,167.8,146.2,144.1]

print('Morphology, Small Radii')
print('------------------------------ \n')

uncert_morph_SR = rms_mean(morph_vals_SR)

print(uncert_morph_SR[0]/morph_vals_SR[1])

print('\n Morphology, Large Radii')
print('------------------------------ \n')

uncert_morph_LSS = rms_mean(morph_vals_LSS)

print(uncert_morph_LSS[0]/morph_vals_LSS[1])

Morphology, Small Radii
------------------------------ 

233.4333333333333
The uncertainty is 11.9, or 5.08% of the mean
0.053474887913771946

 Morphology, Large Radii
------------------------------ 

157.01666666666668
The uncertainty is 10.7, or 6.79% of the mean
0.0683411703143767


* SR: 11.9, or 5.3% of 233
* LSS: 10.7, or 6.8% of 158

## Finally, we obtain the total uncertainty for these three factors. We do so by adding all of these in Quadrature

In [22]:
# small radii
CENTER_VAL_SR = 233.4
CENTER_VAL_LSS = 158.1
M_L_0_UNCERT = 0.083
#print(uncert_mean_SR)
#print(uncert_leuth_SR)
#print(uncert_morph_SR)

total_uncert_SR = mat.sqrt((uncert_mean_SR[0]/CENTER_VAL_SR)**2 + \
                        (uncert_leuth_SR[0]/CENTER_VAL_SR)**2 + \
                        (uncert_morph_SR[0]/CENTER_VAL_SR)**2 + \
                          M_L_0_UNCERT**2)
print('Total Relative Uncertainty for Small Radii: {:.2f}%'\
      .format(total_uncert_SR*100))

total_uncert_LSS = mat.sqrt((uncert_mean_LSS[0]/CENTER_VAL_LSS)**2 + \
                        (uncert_leuth_LSS[0]/CENTER_VAL_LSS)**2 + \
                        (uncert_morph_LSS[0]/CENTER_VAL_LSS)**2 + \
                          M_L_0_UNCERT**2)
print('Total Relative Uncertainty for Large Radii: {:.2f}%'\
      .format(total_uncert_LSS*100))

Total Relative Uncertainty for Small Radii: 13.61%
Total Relative Uncertainty for Large Radii: 14.73%
